### PR TITLE
Clarify reasoning for the use of not_null<T>.

### PR DESCRIPTION
**Pull request by david-mitchell**
Usage of not_null<T> does not mean that null checks are completely unnecessary—it specifies where the responsibility for null checks lies.

### CODE REVIEW

**Comment by blakehawkins** [comment]
Agree that changing this would be useful, but this change seems to make it more ambiguous.

Consider: is not_null something that can be called? Rather, it is a type.

* Suggest rewording and amending the commit to more clearly state who is/is not responsible for the null check.

**Comment by david-mitchell** [comment]
Updated.

**Comment by blakehawkins** [comment]
I think this is pretty good.

**Comment by thecoshman** [comment]
Yes, this a very good update, makes it very clear that checks are still required and where.

**Comment by gdr-at-ms** [comment]
PR has conflict.

**Comment by hsutter** [comment]
Applied separately, thanks! (The item number changed, is now F.23.)
